### ISSUE TITLE
Add ability to run multiple jobs locally in a parallel pool

### ISSUE THREAD

**Issue by cbravo135** [JeffersonLab/hps-mc#153]
We should have a option to run a batch job set via a parallel pool on some local machine.

**Comment by cbravo135** [comment]
```#!/bin/env python

def runCommand(cmd,log=None):
  """
  Provides a wrapper around the subprocess.call command
  cmd should be an array of the form: ["command","arg1",...]
  log should specify a logfile to write to
  """

  import datetime,os,sys
  import subprocess
  from subprocess import CalledProcessError

  try:
    msg = "Executing command:"
    for c in cmd:
      msg+=" %s"%(c)
    sys.stdout.flush()
    returncode = subprocess.call(cmd,stdout=log,stderr=log)
  except CalledProcessError as e:
      msg =  "Caught exception"
      msg+=str(e)
      msg+=" running:"
      for c in cmd:
          msg+=" %s"%(c)
      sys.stdout.flush()
      pass
  return returncode

def launchTests(args):
  return launchTestsArgs(*args)

def launchTestsArgs(tool, cfgFile, fileN, jobN):
  import datetime,os,sys
  import subprocess
  from subprocess import CalledProcessError

  startTime = datetime.datetime.now().strftime("%Y.%m.%d.%H.%M")

  #Build Commands
  setupCmds = []
  preCmd = None
  inDir = "/home/bravo/hps3/mc/mc_2016/wabs/"
  filenameBase = "wabv3-beamv6_2500kBunches_HPS-PhysicsRun2016-Pass2_v4_5_0_pairs1_%i"%fileN
  infilename = inDir+filenameBase+".slcio"
  outDir = "/home/bravo/hps3/users/bravo/mc/wab/"
  outfilename = outDir+filenameBase+".root"
  logfilename = outDir+"logs/"+filenameBase+".log"
  print("%i. Generating %s"%(jobN, outfilename))
  cmd = [ tool, cfgFile, 
          "-i", infilename,
          "-o", outfilename
          ]

  #Execute Commands
  try:
    for setupCmd in setupCmds:
      runCommand(setupCmd)
      pass
    log = file(logfilename,"w")
    if preCmd and config:
      runCommand(preCmd,log)
      pass
    runCommand(cmd,log)
    print("%i. Finished %s"%(jobN, outfilename))
  except CalledProcessError as e:
    print "Caught exception",e
    pass
  return

if __name__ == '__main__':

  import sys,os,signal
  import subprocess
  import itertools
  from multiprocessing import Pool, freeze_support
  from optparse import OptionParser

  parser = OptionParser()
  parser.add_option("-t", "--tool", type="string", dest="tool",
                  help="Tool to use", metavar="tool", default="hpstr")
  parser.add_option("-c", "--config", type="string", dest="configFile",
                  help="Configuration file to use", metavar="configFile", default="dst.py")
  parser.add_option("-f", "--fileList", type="string", dest="fileList",
                  help="List of files to run on.", metavar="fileList", default="fileList.txt")
  parser.add_option("-m", "--fileMod", type="string", dest="fileMod",
                  help="Modifier for output file names.", metavar="fileMod", default="_hpstr")
  parser.add_option("-r", "--rootFile", action="store_true", dest="rootFile",
                  help="Running on root files and not LCIO files", metavar="rootFile")
  parser.add_option("-d", "--debug", action="store_true", dest="debug",
                  help="print extra debugging information", metavar="debug")
  parser.add_option("-p", "--pool", type="int", dest="poolSize",
                  help="Set the number of cores to use (pool size)", metavar="poolSize", default=10)
  parser.add_option("-o", "--outDir", type="string", dest="outDir",
                  help="Specify the directory to save the output", metavar="outDir", default="/data/run/output/")

  (options, args) = parser.parse_args()

  cfgList = []
  fnList = range(1,10001)

  if options.debug:
    print "Testing %i jobs in parallel mode (using Pool(%i))"%(len(fnList),options.poolSize)
    print list(
              itertools.izip([options.tool for x in range(len(fnList))],
                             [options.configFile for x in range(len(fnList))],
                             fnList,
                             fnList
                            )
              )
  else:
    print "Running %i jobs in parallel mode (using Pool(%i))"%(len(fnList),options.poolSize)
    freeze_support()
    # from: https://stackoverflow.com/questions/11312525/catch-ctrlc-sigint-and-exit-multiprocesses-gracefully-in-python
    original_sigint_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
    pool = Pool(options.poolSize)
    signal.signal(signal.SIGINT, original_sigint_handler)
    try:
      res = pool.map_async(launchTests,
                           itertools.izip([options.tool for x in range(len(fnList))],
                                          [options.configFile for x in range(len(fnList))],
                                          fnList,
                                          fnList
                                         )
                           )
      # timeout must be properly set, otherwise tasks will crash
      print res.get(999999999)
      print("Normal termination")
      pool.close()
      pool.join()
    except KeyboardInterrupt:
      print("Caught KeyboardInterrupt, terminating workers")
      pool.terminate()
    except Exception as e:
      print("Caught Exception %s, terminating workers"%(str(e)))
      pool.terminate()
    except: # catch *all* exceptions
      e = sys.exc_info()[0]
      print("Caught non-Python Exception %s"%(e))
      pool.terminate()
```

**Comment by cbravo135** [comment]
That is an example script I commonly use to do this we can base this off of.

**Comment by JeremyMcCormick** [comment]
This would be a nice addition.

The best way would probably be adding an option to this class which currently handles the job execution:

https://github.com/JeffersonLab/hps-mc/blob/master/python/hpsmc/job.py

That way it could be a config switch and existing job scripts would not need to be changed.

**Comment by JeremyMcCormick** [comment]
Resolved by PR #178
